File: neetcode_150/longestSubstringWithoutRepeatingCharacters.java
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
import java.util.HashMap;
import java.util.HashSet;

public class longestSubstringWithoutRepeatingCharacters {
    public static void main(String[] args) {

    }
    public int lengthOfLongestSubstring(String s) {
        if(s.length()<=1){
            return s.length();
        }
        int answer = 0;
        int left = 0, right = 0;
        HashMap<Character, Integer> map = new HashMap<>();
        while(right<s.length()){
            if(map.containsKey(s.charAt(right)) && map.get(s.charAt(right))>=left){
               left = map.get(s.charAt(right))+1;
            }
            map.put(s.charAt(right), right);
            answer = Math.max(right-left+1, answer);
            right++;

        }
        return answer;
    }
}
/* FIRST OWN SOLUTION
    public int lengthOfLongestSubstring(String s) {
        if(s.length()<=1){
            return s.length();
        }
        int answer = 0;
        int left = 0, right = 0;
        HashSet<Character> set = new HashSet<>();
        while(right<s.length()){
            if(!set.contains(s.charAt(right))){
                set.add(s.charAt(right));
                right++;
                answer = Math.max(set.size(),answer);
            }
            else{
                set.remove(s.charAt(left));
                left++;
            }
        }
        return answer;
    }
 */

File: neetcode_150/longestSubstringWithoutRepeatingCharacters.py
from typing import List

def lengthOfLongestSubstring(self, s: str) -> int:
    hs = set()
    left = 0
    for right in range(len(s)):
        while s[right] in hs:
            hs.remove(s[left])
            left+=1
        hs.add(s[right])
        longest = max(longest, right-left+1)
    return longest

# My solution works but the one above is more "pythonic" way to do the same
# def lengthOfLongestSubstring(self, s: str) -> int:
#     if(len(s)<=1):
#         return len(s)
#     hs = set()
#     longest = 1
#     hs.add(s[0])
#     left = 0
#     right = 1
#     while(right<len(s)):
#         if s[right] in hs:
#             longest = max(longest, len(hs))
#             while s[right] in hs:
#                 hs.remove(s[left])
#                 left+=1
#         else:
#             hs.add(s[right])
#             right+=1
#     longest = max(longest, len(hs))
#     return longest










